sdk/database/helper/credsutil/usernames.go

```diff
@@ -0,0 +1,143 @@
+package credsutil
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+type CaseOp int
+
+const (
+	KeepCase CaseOp = iota
+	Uppercase
+	Lowercase
+)
+
+type usernameBuilder struct {
+	displayName string
+	roleName    string
+	separator   string
+
+	maxLen        int
+	caseOperation CaseOp
+}
+
+func (ub usernameBuilder) makeUsername() (string, error) {
+	userUUID, err := RandomAlphaNumeric(20, false)
+	if err != nil {
+		return "", err
+	}
+
+	now := fmt.Sprint(time.Now().Unix())
+
+	parts := []string{
+		"v",
+		ub.displayName,
+		ub.roleName,
+		userUUID,
+		now,
+	}
+	username := joinNonEmpty(ub.separator, parts...)
```

Comment:
Any reason not to have the signature be the same as `strings.Join`, `joinNonEmpty(parts, ub.separator)`?

Reply:
Because you can't put a vararg at the beginning of the signature.

Reply:
You can change the signature to accept a slice

Reply:
Accepting a slice is a lot less flexible than accepting a vararg.

Reply:
Since the separator is a string value, having the input values be passed in as a slice is a bit easier to read IMO

Reply:
Admittedly it's a small issue, but this approach makes it so you don't have to allocate a slice if all of the values are already in variables (or in-line strings). I've put up a commit that shows this (since I neglected to remove the slice from a previous iteration of this function): https://github.com/hashicorp/vault/pull/10050/commits/b6a431d265448c262bef011f4abeb7f5d6bcf28c

Reply:
Yeah, this is purely internal so I'll leave it up to you.